### PR TITLE
aggr-clean

### PR DESCRIPTION
**Pull request by yaronha**
fix aggregator queries in multiple parts and add integration test

### CODE REVIEW

File: pkg/aggregate/iterator.go [hunk 1]
```diff
@@ -176,6 +177,11 @@ func (as *AggregateSeries) NewSetFromChunks(length int) *AggregateSet {
 	for _, aggr := range rawAggregators {
 		if aggr&as.aggrMask != 0 {
 			dataArrays[aggr] = make([]float64, length, length) // TODO: len/capacity & reuse (pool)
+			if aggr == aggrTypeMax || aggr == aggrTypeMin || aggr == aggrTypeLast {
+				for i := 0; i < length; i++ {
+					dataArrays[aggr][i] = math.NaN()
```
Comment (imakhlin) [inline pkg/aggregate/iterator.go]: Note, we use to initialise the data arrays with `(+/-)Inf` rather than `NaN`

File: pkg/aggregate/iterator.go [hunk 2]
```diff
@@ -256,11 +262,11 @@ func (as *AggregateSet) updateCell(aggr AggrType, cell int, val float64) {
 	case aggrTypeSqr:
 		as.dataArrays[aggr][cell] += val * val
 	case aggrTypeMin:
-		if val < as.dataArrays[aggr][cell] {
+		if math.IsNaN(as.dataArrays[aggr][cell]) || val < as.dataArrays[aggr][cell] {
```
Comment (imakhlin) [inline pkg/aggregate/iterator.go]: If you would initialise with proper `Int` value you won't need to change this

File: pkg/querier/querier.go [hunk 3]
```diff
@@ -106,9 +106,13 @@ func (q *V3ioQuerier) selectQry(name, functions string, step int64, windows []in
 func (q *V3ioQuerier) queryNumericPartition(
 	partition *partmgr.DBPartition, name, functions string, step int64, windows []int, filter string) (SeriesSet, error) {
 
-	mint, maxt := partition.GetPartitionRange(q.maxt)
+	mint, maxt := partition.GetPartitionRange()
 	if q.mint > mint {
 		mint = q.mint
+		if step != 0 {
+			// temporary Aggregation fix: if mint is not aligned with steps, move it to the next step tick
+			mint += (maxt - q.mint) % step
```
Comment (imakhlin) [inline pkg/querier/querier.go]: you can round up with `mint += step-mint%step`
and round down with `mint -= mint%step`

but I think it is expected that reminder interval should also be counted.

File: test/aggregates/aggregates_test.go [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		if err != nil {
+			t.Fatal(err)
+		}
+		index = (index + 1) % len(testConfig.Values)
+		total++
+	}
+	fmt.Println("total samples written:", total)
+
+	appender.WaitForCompletion(0)
+	time.Sleep(time.Second * 2)
```
Comment (imakhlin) [inline test/aggregates/aggregates_test.go]: Why sleep?

File: test/aggregates/aggregates_test.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+			}
+
+		}
+		if iter.Err() != nil {
+			t.Fatal(iter.Err())
+		}
+
+		if aggr == "count" {
+			fmt.Println("\nTotal count read:", totalCount)
+			if totalCount != total {
```
Comment (imakhlin) [inline test/aggregates/aggregates_test.go]: use Go test frameworks like testify, see `github.com/stretchr/testify/assert`

File: test/aggregates/aggregates_test.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+type metricContext struct {
+	lset utils.Labels
+	ref  uint64
+}
+
+func TestAggregates(t *testing.T) {
+
+	testConfig := TestConfig{
```
Comment (imakhlin) [inline test/aggregates/aggregates_test.go]: I would recommend separating the test onto stages:
1. Setup - prepare the data and prerequisites required for running the test like table with desired partitioning, etc
2. TearDown - cleanup after test (could be skipped if test has failed to make investigation possible)
3. The test code itself.

Nice example: `pkg/aggregate/aggregate_test.go`

Comment (imakhlin) [reply test/aggregates/aggregates_test.go]: This integration test wont pass twice.

File: test/aggregates/aggregates_test.go [hunk 7]
```diff
@@ -0,0 +1,208 @@
+// +build integration
```
Comment (gtopper) [inline test/aggregates/aggregates_test.go]: Should rename file accordingly (`aggregates_integration_test`).

File: pkg/aggregate/iterator.go (hunk 1 above)
Comment (yaronha) [reply pkg/aggregate/iterator.go]: @imakhlin as discussed i think its simpler/cleaner (not confined to expr constraints), u can change if to the other way if u think its critical

File: pkg/aggregate/iterator.go (hunk 2 above)
Comment (yaronha) [reply pkg/aggregate/iterator.go]: see above

File: pkg/querier/querier.go (hunk 3 above)
Comment (yaronha) [reply pkg/querier/querier.go]: @imakhlin expected return is results with time > query mint, with your suggestion the response will not comply with the user request

File: test/aggregates/aggregates_test.go [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+implied. See the License for the specific language governing
+permissions and limitations under the License.
+
+In addition, you may not use the software for any purposes that are
+illegal under applicable law, and the grant of the foregoing license
+under the Apache 2.0 license is conditioned upon your compliance with
+such restriction.
+*/
+
+package aggregates
```
[the remark this answers is not in the document]
Comment (yaronha) [reply test/aggregates/aggregates_test.go]: @imakhlin  ok, please place it in there and can also use the create/tear-down flow

**Comment by gshatz** [comment]
I am merging this pr although there are still some changes to be done in the integration tests due to delivery constraints

**Pull request merged**